### ISSUE TITLE
System.IO.IOException

### ISSUE THREAD

**Issue by TBag** [pnp/pnpjs#692]
### Category
- [ ] Enhancement
- [x] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version
Please specify what version of the library you are using: [ 1.3.2 ]
Please specify what version(s) of SharePoint you are targeting: [ Office 365 ]

### Expected / Desired Behavior / Question
When running in node.js, we are expecting to be able to download a file from a SharePoint site, then upload the file to another SharePoint site with this code.

web.getFileByServerRelativeUrl("/sites/sitename1/Templates/file.png").getBuffer().then(buffer => {
    web.getFolderByServerRelativeUrl("/sites/sitename2/Templates").files.add("new.png", buffer, true).then(fileAddResult =>{
        console.log('File was uploaded successfully!')
    })
})

We have also tried copying the file instead of downloading and uploading with this code.

web.getFileByServerRelativeUrl("/sites/sitename1/Templates/file.png")
    .copyTo("/sites/sitename2/Shared%20Documents/-new.png")

### Observed Behavior

This error occurs:

(node:8312) UnhandledPromiseRejectionWarning: Error: Error making HttpClient request in queryable [500] Internal Server Error ::> {"odata.error":{"code":"-1, System.IO.IOException","message":{"lang":"en-US","value":"I/O error occurred."}}}
    at new HttpRequestError (C:\EMS\nodejs-webapp\node_modules\@pnp\odata\dist\odata.es5.umd.js:96:32)
    at C:\EMS\nodejs-webapp\node_modules\@pnp\odata\dist\odata.es5.umd.js:105:24
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:189:7)

### Steps to Reproduce

Download the files in [this repo](https://github.com/TBag/pnp-sp-issue).

Modify the values in Settings.ts for your environment.

Modify lines 24, 26, and 27 in app.ts to match your tenant and an existing SharePoint site, library and file.

Execute these commands:

npm install
tsc
node app.js

Then you will see the error message.

### Workaround

Uncomment line 2 and line 3 in app.ts, they look like this.

import { patch } from './patch'
patch();

Here are the contents of patch.ts.

declare var global: any;
const NodeFetch = require("node-fetch");

export function patch() {

    // patch these globally for nodejs
    if (!global.Headers) {
        global.Headers = NodeFetch.Headers;
    }
    if (!global.Request) {
        global.Request = NodeFetch.Request;
    }
    if (!global.Response) {
        global.Response = NodeFetch.Response;
    }
    if (!global.fetch) {
        global.fetch = NodeFetch;
    }
}

Execute these commands:

tsc
node app.js

Then you will see the success message.

###Additional Information

This issue is similar to issue #114.  However, the workaround in issue #114 does not fix this issue.

First, we don’t think the headers should be in the sp.setup, because sp.setup is for the entity app and should be invoked only once.

sp.setup({
    sp: {
        fetchClientFactory: () => {
            return new NodeFetchClient(spCredentials, siteUrl);
        },
        headers: {
            "Accept": "application/json; odata=verbose",
            "Content-Length": fileData.length
        }
    },
});

We think the magic happens in the NodeFetchClient. It also uses node-fetch, but only the global fetch is patched. 

We tried updating our AdalCertFetchClient following this approach and added the following code to AdalCertFetchClient.ts:

import { default as fetch } from "node-fetch";

Then we got the error: Error making HttpClient request inqueryable [403] Forbidden

We think this is because @microsoft/microsoft-graph-client patches the global Header with a different library.

If you comment out line 11 in the app.ts:

// GraphClient.initWithMiddleware({ authProvider: null})

Then execute these commands:

tsc
node app.js

You will get the error: ReferenceError: Headers is not defined

@pnp/nodejs also patches these globally: 

https://github.com/pnp/pnpjs/blob/dev/packages/nodejs/src/nodejs.ts. 

But we are not sure when and where the patch would apply.

Instead of letting the libraries (@pnp and @microsoft/microsoft-graph-client) patch these globally differently and uncontrollably, we patch them in advance.  This is how we created the workaround. 

Note: The libraries check before patching.  So after we patch, they will not patch again.

**Comment by TBag** [comment]
Hi @patrick-rodgers,

Here's some additional information about this issue.

We tried to bundle the Node.js app with webpack.  But we found our patch does not work after the app is bundled.  It seems quite hard to make these libraries and webpack work together.  As a workaround we use this process to package the app.

Run these commands

```
npm run build
cd dist
npm install --production
```

Then, zip the dist folder.

**Comment by patrick-rodgers** [comment]
Going to close this as I am not sure what we can do to help here. You have a successful workaround and our patching is gated so we aren't double patching or interfering once things are already patched. I don't have any control over what the graph SDK team does and they aren't interested in my feedback. We patch the headers when the [imports are done from the nodejs library](https://github.com/pnp/pnpjs/blob/dev/packages/nodejs/src/nodejs.ts#L5) so perhaps there is a way to import our code first or second depending on what you are attempting to accomplish.

If this is a blocking issue or you have a suggestion on what you think we should change please open an new issue stating that and reference this one and we will take a look.

As for the headers you can define them in setup or using the [configure method](https://pnp.github.io/pnpjs/odata/docs/queryable/#configure) on specific instances.

There was a bug that was fixed so that [all clients use the same fetch method](https://github.com/pnp/pnpjs/pull/770), perhaps that helps you as well.
